File: packages/python/examples/bsc_x402_sign_typed_data.py
"""
Demo: Sign EIP-712 typed data for x402 payment permit with an EVM/BSC wallet.

This example is the EVM/BSC counterpart to tron_x402_sign_typed_data.py.
It resolves the active wallet from environment variables via WalletFactory()
and verifies the recovered signer directly against the EVM address.

Recommended env:
  EVM_PRIVATE_KEY=<hex> python examples/bsc_x402_sign_typed_data.py
  EVM_MNEMONIC="<words>" python examples/bsc_x402_sign_typed_data.py

Optional local mode also works:
  AGENT_WALLET_PASSWORD=<password> python examples/bsc_x402_sign_typed_data.py
"""

import asyncio

from eth_account import Account
from eth_account.messages import encode_typed_data

from agent_wallet import WalletFactory

# --- x402 PaymentPermit typed data (BSC example) ---

PAYMENT_PERMIT = {
    "types": {
        "EIP712Domain": [
            {"name": "name", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
        ],
        "PaymentPermitDetails": [
            {"name": "buyer", "type": "address"},
            {"name": "amount", "type": "uint256"},
            {"name": "nonce", "type": "uint256"},
        ],
    },
    "primaryType": "PaymentPermitDetails",
    "domain": {
        "name": "x402PaymentPermit",
        "chainId": 97,  # BSC testnet
        "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
    },
    "message": {
        "buyer": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
        "amount": 1000000,
        "nonce": 0,
    },
}

STANDARD_TYPED_DATA = {
    "types": {
        "EIP712Domain": [
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
        ],
        "Transfer": [
            {"name": "to", "type": "address"},
            {"name": "amount", "type": "uint256"},
        ],
    },
    "primaryType": "Transfer",
    "domain": {
        "name": "MyDApp",
        "version": "1",
        "chainId": 97,
        "verifyingContract": "0x1234567890abcdef1234567890abcdef12345678",
    },
    "message": {
        "to": "0xAbCdEf0123456789AbCdEf0123456789AbCdEf01",
        "amount": 5000000,
    },
}


async def main():
    provider = WalletFactory()
    wallet = await provider.get_active_wallet()
    address = await wallet.get_address()

    if not address.startswith("0x"):
        raise RuntimeError(
            "bsc_x402_sign_typed_data.py expects an EVM wallet. "
            "Set EVM_PRIVATE_KEY or EVM_MNEMONIC."
        )

    print(f"Address: {address}")
    print()

    print("=== x402 PaymentPermit (BSC) ===")
    print(f"  Domain:      {PAYMENT_PERMIT['domain']['name']}")
    print(f"  Chain ID:    {PAYMENT_PERMIT['domain']['chainId']}")
    print(f"  Buyer:       {PAYMENT_PERMIT['message']['buyer']}")
    print(f"  Amount:      {PAYMENT_PERMIT['message']['amount']}")

    sig1 = await wallet.sign_typed_data(PAYMENT_PERMIT)
    print(f"  Signature:   {sig1}")
    print()

    print("=== Standard EIP-712 Transfer ===")
    print(
        f"  Domain:      {STANDARD_TYPED_DATA['domain']['name']} "
        f"v{STANDARD_TYPED_DATA['domain']['version']}"
    )
    print(f"  To:          {STANDARD_TYPED_DATA['message']['to']}")
    print(f"  Amount:      {STANDARD_TYPED_DATA['message']['amount']}")

    sig2 = await wallet.sign_typed_data(STANDARD_TYPED_DATA)
    print(f"  Signature:   {sig2}")
    print()

    print("=== Verify Signature ===")
    signable = encode_typed_data(full_message=PAYMENT_PERMIT)
    recovered = Account.recover_message(signable, signature=bytes.fromhex(sig1))
    print(f"  Recovered:   {recovered}")
    print(f"  Matches:     {recovered.lower() == address.lower()}")


if __name__ == "__main__":
    asyncio.run(main())

File: packages/typescript/examples/bsc-x402-sign-typed-data.ts
/**
 * Demo: Sign EIP-712 typed data for x402 payment permit with an EVM/BSC wallet.
 *
 * This example is the EVM/BSC counterpart to tron-x402-sign-typed-data.ts.
 * It resolves the active wallet from environment variables via resolveWalletProvider()
 * and verifies the recovered signer directly against the EVM address.
 *
 * Recommended env:
 *   AGENT_WALLET_PRIVATE_KEY=<hex> npx tsx examples/bsc-x402-sign-typed-data.ts
 *   AGENT_WALLET_MNEMONIC="word1 word2 ..." npx tsx examples/bsc-x402-sign-typed-data.ts
 *
 * Optional local mode also works:
 *   AGENT_WALLET_PASSWORD=<password> npx tsx examples/bsc-x402-sign-typed-data.ts
 */

import { resolveWalletProvider, type Eip712Capable } from "../src/index.js";
import { recoverTypedDataAddress } from "viem";

const PAYMENT_PERMIT = {
  types: {
    EIP712Domain: [
      { name: "name", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" },
    ],
    PaymentPermitDetails: [
      { name: "buyer", type: "address" },
      { name: "amount", type: "uint256" },
      { name: "nonce", type: "uint256" },
    ],
  },
  primaryType: "PaymentPermitDetails",
  domain: {
    name: "x402PaymentPermit",
    chainId: 97,
    verifyingContract: "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
  },
  message: {
    buyer: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
    amount: 1000000,
    nonce: 0,
  },
};

const STANDARD_TYPED_DATA = {
  types: {
    EIP712Domain: [
      { name: "name", type: "string" },
      { name: "version", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" },
    ],
    Transfer: [
      { name: "to", type: "address" },
      { name: "amount", type: "uint256" },
    ],
  },
  primaryType: "Transfer",
  domain: {
    name: "MyDApp",
    version: "1",
    chainId: 97,
    verifyingContract: "0x1234567890AbcdEF1234567890aBcdef12345678",
  },
  message: {
    to: "0xabCDeF0123456789AbcdEf0123456789aBCDEF01",
    amount: 5000000,
  },
};

async function main() {
  const provider = resolveWalletProvider({ network: "eip155:97" });
  const wallet = await provider.getActiveWallet();
  const address = await wallet.getAddress();

  if (!address.startsWith("0x")) {
    throw new Error(
      "bsc-x402-sign-typed-data.ts expects an EVM wallet. Set AGENT_WALLET_PRIVATE_KEY or AGENT_WALLET_MNEMONIC.",
    );
  }

  console.log(`Address: ${address}`);
  console.log();

  const signer = wallet as unknown as Eip712Capable;

  console.log("=== x402 PaymentPermit (BSC) ===");
  console.log(`  Domain:      ${PAYMENT_PERMIT.domain.name}`);
  console.log(`  Chain ID:    ${PAYMENT_PERMIT.domain.chainId}`);
  console.log(`  Buyer:       ${PAYMENT_PERMIT.message.buyer}`);
  console.log(`  Amount:      ${PAYMENT_PERMIT.message.amount}`);

  const sig1 = await signer.signTypedData(PAYMENT_PERMIT);
  console.log(`  Signature:   ${sig1}`);
  console.log();

  console.log("=== Standard EIP-712 Transfer ===");
  console.log(
    `  Domain:      ${STANDARD_TYPED_DATA.domain.name} v${STANDARD_TYPED_DATA.domain.version}`,
  );
  console.log(`  To:          ${STANDARD_TYPED_DATA.message.to}`);
  console.log(`  Amount:      ${STANDARD_TYPED_DATA.message.amount}`);

  const sig2 = await signer.signTypedData(STANDARD_TYPED_DATA);
  console.log(`  Signature:   ${sig2}`);
  console.log();

  console.log("=== Verify Signature ===");
  const { EIP712Domain, ...msgTypes } = PAYMENT_PERMIT.types;
  const recovered = await recoverTypedDataAddress({
    domain: PAYMENT_PERMIT.domain as any,
    types: msgTypes as any,
    primaryType: PAYMENT_PERMIT.primaryType,
    message: PAYMENT_PERMIT.message as any,
    signature: `0x${sig1}`,
  });

  console.log(`  Recovered:   ${recovered}`);
  console.log(`  Matches:     ${recovered.toLowerCase() === address.toLowerCase()}`);
}

main().catch(console.error);
